File: aoc/2024/aoc202419.hs
module AOC202419 where

import Data.Set(Set,fromList,member)

import AOC

aoc = AOC {
    day="19",
    aocTests=[
        AOCTest {
            testData=unlines [
                "r, wr, b, g, bwu, rb, gb, br",
                "",
                "brwrr",
                "bggr",
                "gbbr",
                "rrbgbr",
                "ubwu",
                "bwurrg",
                "brgr",
                "bbrgwb"
                ],
            testResult=Just "6",
            testResult2=Just "16"
            }
        ],
    aocCode=ParallelCode {
        pcodeParse=const parse,
        pcodeParse2=const parse,
        pcodeTest=const result,
        pcodeTest2=result2,
        pcodeResult=const result,
        pcodeResult2=result2
        }
    }

maxLen = 8 -- hard coded maximum pattern length

parse :: String -> (Set String,[String])
parse = p . lines
  where p (patterns:"":designs) =
            (fromList $ words $ filter (/= ',') patterns,designs)

possible :: Set String -> String -> Bool
possible patterns design = evalMemoized (possibleM design)
  where
    possibleM = memoize p
    p design
      | null design = return True
      | otherwise = check [1..maxLen]
      where
        check [] = return False
        check (n:ns)
          | take n design `member` patterns = do
              res <- possibleM (drop n design)
              if res
                then return True
                else check ns
          | otherwise = check ns

result (patterns,designs) = length $ filter (possible patterns) designs

ways :: Set String -> String -> Int
ways patterns design = evalMemoized (waysM design)
  where
    waysM = memoize w
    w design
      | null design = return 1
      | otherwise = do
          results <- mapM check [1..min maxLen (length design)]
          return $ sum results
      where
        check n
          | take n design `member` patterns = waysM $ drop n design
          | otherwise = return 0

result2 ncpu (patterns,designs) =
    parallelMapReduce ncpu (ways patterns) sum designs

File: aoc/2025/aoc202419.go
package main

func init() {
	Register(&aoc202419{
		AOC: AOC{
			Day:           19,
			InputFilename: "../2024/input/19.txt",
			Tests: []Test{
				Test{
					Input: `r, wr, b, g, bwu, rb, gb, br

brwrr
bggr
gbbr
rrbgbr
ubwu
bwurrg
brgr
bbrgwb
`,
					Part1: "6",
					Part2: "16",
				},
			},
		},
	})
}

type aoc202419 struct {
	AOC
}

func (aoc *aoc202419) patterns(input *Input) []string {
	paragraph, _ := input.Paragraph()
	input = InputString(paragraph)
	pats := []string{}
	for pat, ok := input.Word(); ok; pat, ok = input.Word() {
		if pat[len(pat)-1] == ',' {
			pat = pat[:len(pat)-1]
		}
		pats = append(pats, pat)
	}
	return pats
}

func (aoc *aoc202419) possible(design string, pats []string, memo map[string]bool) bool {
	if p, ok := memo[design]; ok {
		return p
	}
	for _, pat := range pats {
		if len(pat) > len(design) {
			continue
		}
		if pat == design[:len(pat)] {
			d := design[len(pat):]
			if d == "" || aoc.possible(d, pats, memo) {
				memo[design] = true
				return true
			}
		}
	}
	memo[design] = false
	return false
}

func (aoc *aoc202419) Part1(input *Input) string {
	pats := aoc.patterns(input)
	memo := map[string]bool{}
	result := 0
	for design, ok := input.Line(); ok; design, ok = input.Line() {
		if aoc.possible(design, pats, memo) {
			result++
		}
	}
	return IntResult(result)
}

func (aoc *aoc202419) countArrangements(design string, pats []string, memo map[string]int) int {
	if n, ok := memo[design]; ok {
		return n
	}
	n := 0
	for _, pat := range pats {
		if len(pat) > len(design) {
			continue
		}
		if pat == design[:len(pat)] {
			d := design[len(pat):]
			if d == "" {
				n++
			} else {
				n += aoc.countArrangements(d, pats, memo)
			}
		}
	}
	memo[design] = n
	return n
}

func (aoc *aoc202419) Part2(input *Input) string {
	pats := aoc.patterns(input)
	memo := map[string]int{}
	result := 0
	for design, ok := input.Line(); ok; design, ok = input.Line() {
		result += aoc.countArrangements(design, pats, memo)
	}
	return IntResult(result)
}
